_posts/fr/2018-02-11-throwable-error-exception.md

```diff
@@ -0,0 +1,374 @@
+Mais aussi pour éviter d'afficher des informations sensibles (url du fichier, stack trace, message d'erreur ...)
+La bonne pratique étant de ne pas laisser une exception casser le site._
+
+```
+set_exception_handler(function($exception){
+    echo 'Une erreur est survenue. Veuillez rééssayer ulterieurement.';
+    // log($exception->getMessage());
+    // email au developpeur
+});
+```
+
+### Les codes d'erreurs
+
+Le code d'erreur est un `integer` qui peut être utilisé pour codifier/identifier l'erreur.
+
+> Il permet par exemple d'afficher le code de l'erreur plutôt que le message de l'`Exception` au visiteur. 
+> Il permet de masquer la raison de l'erreur, qui peut dans certains cas contenir des informations sensibles. 
+
+## Utilisation avancée
+
+### Créer une exception personalisée
+
+Il est très utile de créer des exceptions personnalisées afin qu'elles puissent identifier un problème plus précisément, mais aussi afin de pouvoir transporter des données supplémentaires (texte, object, array...).
+
+```php
+class MyObject
+{
+    public $content;
+}
+
+class MyObjectException extends RuntimeException
+{
+    /**
+     * @var MyObject
+     */
+    private $myObject;
+
+    public function __construct(MyObject $myObject, $message = "", $code = 0, Throwable $previous = null)
+    {
+        parent::__construct($message, $code, $previous);
+        $this->myObject = $myObject;
+    }
+
+    /**
+     * @return MyObject
+     */
+    public function getMyObject()
+    {
+        return $this->myObject;
+    }
+}
+```
+
+> Quand l'`MyObjectException` est attrapé, on peut récupérer l'objet `MyObject` via la méthode `getMyObject()`
+> ce qui permet de gérer encore plus précisément la gestion d'erreur. 
+
+### Relancer une exception
+
+Parfois il est utile de tracer/logguer ce qui s'est mal déroulé.
+Dans ce cas on va donc attraper l'`Exception`, logger un message d'erreur avant de relancer l'`Exception`.
+
+```php
+try {
+    // mise à jour d'un contenu
+}
+catch (Exception $e) {
+    // log('La mise à jour a échoué.');
+    throw $e;
+}
+```
+
+Exemple concret :
+
+```php
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Psr\Log\NullLogger;
+
+class UserFactory implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+    
+    private $passwordGenerator;
+    
+    public function construct(PasswordGeneratorInterface $passwordGenerator)
+    {
+        $this->passwordGenerator = $passwordGenerator;
+        $this->logger = new NullLogger();
+    }
+    
+    public function create() 
+    {
+        try {
+            $user = new User();
+            $password = $this->passwordGenerator->generatePassword();
+            $user->setPassword($password);
+            
+            return $user;
+        } catch (Exception $exception) {
+            $this->logger->error('Une erreur est survenue pendant la creation d'un utilisateur. Raison: ' . $exception->getMessage());
```

Comment:
`$this->logger->error('Une erreur est survenue pendant la creation d\'un utilisateur. Raison: {exceptionMessage}.', ['exceptionMessage' => $exception->getMessage()]);`

Reply:
my goal is not to explain the monolog usage in this article
i'll take the `\'` correction

Reply:
This is not monolog linked, its PSR 🍞